### PR TITLE
feat: enrich token swap transactions

### LINKED ISSUE #3083: Enrich Transactions - Display Tokens Exchange
When a transaction takes place through token swap we need to display the token exchange in a similar format like DEX Transactions. 

We need to display additionally the fee incurred in the transaction.

### PR DESCRIPTION
**Pull request by CedrikNikita**
fixes #3083

### CODE REVIEW

File: src/composables/fungibleTokens.ts [hunk 1]
```diff
@@ -106,11 +109,36 @@ export function useFungibleTokens() {
...
     getLastActiveProtocolAccount,
   } = useAccounts();
+  const { getCurrentCurrencyRate } = useCurrencies();
 
   function getAccountTokenBalances(address: AccountAddress): ITokenBalance[] {
     return tokenBalances.value.filter((token) => token.address === address) || [];
   }
 
+  const accountsTotalTokenBalance = computed(() => {
+    const result = accounts.value.reduce(
```
Comment (peronczyk) [inline src/composables/fungibleTokens.ts:119]: You can return the result immediatelly:
```suggestion
  const accountsTotalTokenBalance = computed(
    () => accounts.value
      .reduce((total, account) => {
        ...
      })
      .toFixed(2),
  );
```

File: src/composables/transactionData.ts [hunk 2]
```diff
@@ -109,6 +116,25 @@ export function useTransactionData({
     ),
   );
 
+  const isTokenSwap = computed(
+    (): boolean => tokenContractAddresses.value
+      .concat(tokenSwapAddresses.value).some((address) => address === innerTx.value?.contractId),
```
Comment (peronczyk) [inline src/composables/transactionData.ts:121]: I think this is more readable:
```suggestion
    (): boolean => [...tokenContractAddresses.value, ...tokenSwapAddresses.value].some((address) => address === innerTx.value?.contractId),
```

File: src/popup/components/SwapRates.vue [hunk 3]
```diff
@@ -62,13 +63,14 @@ export default defineComponent({
...
+    const { tokenSwapAddressToTokenContractAddress } = useAeTokenSwaps();
 
-    const { isDexSwap, txFunctionParsed } = useTransactionData({
+    const { isDexSwap, txFunctionParsed, isTokenSwap } = useTransactionData({
       transaction: toRef(() => props.transaction),
     });
 
     const rates = computed(() => {
-      if (!isDexSwap.value || !txFunctionParsed.value) {
+      if (!(isDexSwap.value || isTokenSwap.value) || !txFunctionParsed.value) {
```
Comment (peronczyk) [inline src/popup/components/SwapRates.vue:73]: Isn't this equal to:
```suggestion
      if (!isDexSwap.value || !isTokenSwap.value || !txFunctionParsed.value) {
```

File: src/popup/pages/Assets/AssetDetailsInfo.vue [hunk 4]
```diff
@@ -58,9 +97,14 @@
...
-        />
-
+        >
+          <template #text>
+            <TokenAmount
+              :amount="assetData.marketCap"
+              :protocol="assetData.protocol"
+            />
+          </template>
+        </DetailsRow>
```
Comment (peronczyk) [inline src/popup/pages/Assets/AssetDetailsInfo.vue:107]: I suggest to add empty line after this one.

File: src/protocols/aeternity/composables/aeTokenSwaps.ts [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+const POLLING_INTERVAL = 60000;
+
+const initPollingWatcher = createPollingBasedOnMountedComponents(POLLING_INTERVAL);
+
+/**
+ * List of all token swap available
+ */
+const tokenSwaps = ref<ITokenSwap[]>([]);
+
+export function useAeTokenSwaps() {
```
Comment (peronczyk) [inline src/protocols/aeternity/composables/aeTokenSwaps.ts:26]: Maybe you can use `createCustomScopedComposable` here?

File: src/protocols/aeternity/composables/aeTokenSwaps.ts [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+import { AE_TOKEN_SWAPS_URLS } from '@/protocols/aeternity/config';
+import { createPollingBasedOnMountedComponents } from '@/composables/composablesHelpers';
+
+let composableInitialized = false;
+
+const areTokenSwapsReady = ref(false);
+
+const tokenSwapsUrl = ref('');
+
+const POLLING_INTERVAL = 60000;
```
Comment (peronczyk) [inline src/protocols/aeternity/composables/aeTokenSwaps.ts:17]: I suggets to put the constants at the beginning and then states (avoid mixing them).

File: src/protocols/aeternity/composables/aeTokenSwaps.ts [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+  function tokenSwapAddressToTokenContractAddress(
+    address: string,
+  ): string {
+    const tokenSwap = tokenSwaps.value.find((sale) => sale.saleAddress === address);
+    return tokenSwap?.address ?? address;
+  }
+
+  async function fetchTokenSwaps(path: string) {
+    const res = await fetchJson(`${tokenSwapsUrl.value}/${path}`);
+    const meta = res?.meta;
```
Comment (peronczyk) [inline src/protocols/aeternity/composables/aeTokenSwaps.ts:45]:
```suggestion
    const { items, meta } = res || {};
```

File: src/protocols/aeternity/composables/aeTokenSwaps.ts [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    return {
+      data: res?.items,
+      meta,
+      next: meta?.totalPages && meta.totalPages > meta?.currentPage
```
Comment (peronczyk) [inline src/protocols/aeternity/composables/aeTokenSwaps.ts:49]: The `meta` could be `undefined` so `meta.totalPages` can cause errors.

File: src/protocols/aeternity/libs/AeternityAdapter.ts [hunk 9]
```diff
@@ -259,17 +262,27 @@ export class AeternityAdapter extends BaseProtocolAdapter {
...
         () => fetchFromMiddleware(`/v2/aex9/account-balances/${address}?limit=100`),
         fetchFromMiddleware,
       ));
+      if (tokens.length) {
+        await Promise.race(
+          [
+            watchUntilTruthy(areTokenSwapsReady),
+            sleep(5000),
+          ],
+        );
```
Comment (peronczyk) [inline src/protocols/aeternity/libs/AeternityAdapter.ts:277]:
```suggestion
        await Promise.race([
          watchUntilTruthy(areTokenSwapsReady),
          sleep(5000),
        ]);
```

**Review by peronczyk — APPROVED** [review]

**Comment by Liubov-crypto** [comment]
LGTM

**Comment by Liubov-crypto** [comment]
Fixed, LGTM

**Pull request merged**